Mage.Sets/src/mage/cards/k/KaghaShadowArchdruid.java

```diff
@@ -0,0 +1,168 @@
+import mage.abilities.effects.AsThoughEffectImpl;
+import mage.abilities.effects.common.MillCardsControllerEffect;
+import mage.abilities.effects.common.continuous.GainAbilitySourceEffect;
+import mage.abilities.keyword.DeathtouchAbility;
+import mage.constants.SubType;
+import mage.constants.SuperType;
+import mage.constants.WatcherScope;
+import mage.constants.Zone;
+import mage.filter.FilterCard;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.events.ZoneChangeEvent;
+import mage.game.permanent.Permanent;
+import mage.watchers.Watcher;
+import mage.cards.Card;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.AsThoughEffectType;
+import mage.constants.CardType;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+
+/**
+ *
+ * @author Fyusel
+ */
+public final class KaghaShadowArchdruid extends CardImpl {
+
+    private static final FilterCard filter = new FilterCard(
+            "you may play a land or cast a permanent spell from among cards in your graveyard that were put there from your library this turn"
+    );
+
+    static {
+        filter.add(KaghaShadowArchdruidWatcher::checkCard);
+    }
+
+    public KaghaShadowArchdruid(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{2}{B}{G}");
+        
+        this.supertype.add(SuperType.LEGENDARY);
+        this.subtype.add(SubType.ELF);
+        this.subtype.add(SubType.DRUID);
+        this.power = new MageInt(1);
+        this.toughness = new MageInt(4);
+
+        // Whenever Kagha, Shadow Archdruid attacks, it gains deathtouch until end of turn. Mill two cards.
+        Ability ability = new AttacksTriggeredAbility(new GainAbilitySourceEffect(
+            DeathtouchAbility.getInstance(),
+            Duration.EndOfTurn));
+        ability.addEffect(new MillCardsControllerEffect(2));
+        this.addAbility(ability);
+
+        // Once during each of your turns, you may play a land or cast a permanent spell from among cards in your graveyard that were put there from your library this turn.
+        Ability kaghaSpecialAbility = new SimpleStaticAbility(Zone.BATTLEFIELD,
+            new KaghaShadowArchdruidEffect(filter, "Once during each of your turns, you may play a land or cast a permanent spell from among cards in your graveyard that were put there from your library this turn."))
+            .setIdentifier(MageIdentifier.KaghaShadowArchdruidWatcher
+        );
+        this.addAbility(kaghaSpecialAbility, new KaghaShadowArchdruidWatcher());
+    }
+
+    private KaghaShadowArchdruid(final KaghaShadowArchdruid card) {
+        super(card);
+    }
+
+    @Override
+    public KaghaShadowArchdruid copy() {
+        return new KaghaShadowArchdruid(this);
+    }
+}
+
+class KaghaShadowArchdruidEffect extends AsThoughEffectImpl {
+
+    private final FilterCard filter;
+
+    KaghaShadowArchdruidEffect(FilterCard filter, String text) {
```

Comment:
Rather than pass in parameters to a custom effect, just set them in the constructor. I also don't think it makes much sense to create a filter object at all; just check the relevant info directly in the `applies()` method.

Reply:
Done